src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -970,17 +972,22 @@ private static ImmutableArray<NamedTypeSymbol> GetBaseInterfaces(NamedTypeSymbol
 
             return result.ToImmutableAndFree();
 
-            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved)
+            static void addAllInterfaces(NamedTypeSymbol @interface, HashSet<NamedTypeSymbol> visited, ArrayBuilder<NamedTypeSymbol> result, ConsList<TypeSymbol> basesBeingResolved, ImmutableHashSet<NamedTypeSymbol> cycleGuard)
             {
-                if (@interface.IsInterface && visited.Add(@interface))
+                if (@interface.IsInterface && !cycleGuard.Contains(@interface.OriginalDefinition) && visited.Add(@interface))
```

Comment:
>cycleGuard.Contains(@interface.OriginalDefinition) [](start = 47, length = 50)

I have a couple of issues with this approach:
- It is not documented or implemented in a way that relates it to the language specification (underlying or feature's language spec). This leaves the impression to the reader that it is an ad-hoc solution to a perceived problem with the compiler's behavior. This could be fixed by modeling it on the method whose behavior it is trying to simulate (i.e. using `BaseTypeAnalysis.TypeDependsOn` as in `SourceNamedTypeSymbol.MakeAcyclicInterfaces`). Alternatively, if we must have a distinct technique to accomplish the same thing in two different places in the compiler, at least add a comment explaining how this relates to the specification and how it relates to the method whose behavior it is trying to simulate.
- It uses a completely different technique than that used for `basesBeingResolved` (`ImmutableHashSet` instead of `ConsList`) for no particularly good reason, making the compiler harder to maintain.
- Although it performs the check in terms of original definitions, it does not use reference equality for the check but rather type equality.

Reply:
@gafter

> I have a couple of issues with this approach:

I think this is now addressed as we discussed offline.

---
In reply to: [325870833](https://github.com/dotnet/roslyn/pull/38738#discussion_r325870833) [](ancestors = 325870833)

Reply:
Please use `cycleGuard.ContainsReference`.  It is an extension method used everywhere we use `basesBeingResolved` to check for cycles.

---
In reply to: [325893335](https://github.com/dotnet/roslyn/pull/38738#discussion_r325893335) [](ancestors = 325893335,325870833)

Reply:
OOps, you already do.

---
In reply to: [325896204](https://github.com/dotnet/roslyn/pull/38738#discussion_r325896204) [](ancestors = 325896204,325893335,325870833)